packages/core/strapi/lib/commands/builders/tasks/dts.js

```diff
@@ -0,0 +1,191 @@
+  const compilerOptions = {
+    ...tsconfig.options,
+    declaration: true,
+    declarationDir: outDir,
+    emitDeclarationOnly: true,
+    noEmit: false,
+    outDir,
+  };
+
+  const program = ts.createProgram(tsconfig.fileNames, compilerOptions);
+
+  const emitResult = program.emit();
+
+  const allDiagnostics = ts.getPreEmitDiagnostics(program).concat(emitResult.diagnostics);
+
+  for (const diagnostic of allDiagnostics) {
+    if (diagnostic.file && diagnostic.start) {
+      const { line, character } = ts.getLineAndCharacterOfPosition(
+        diagnostic.file,
+        diagnostic.start
+      );
+      const message = ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n');
+
+      const file = path.relative(cwd, diagnostic.file.fileName);
+
+      const output = [
+        `${chalk.cyan(file)}:${chalk.cyan(line + 1)}:${chalk.cyan(character + 1)} - `,
+        `${chalk.gray(`TS${diagnostic.code}:`)} ${message}`,
+      ].join('');
+
+      if (diagnostic.category === ts.DiagnosticCategory.Error) {
+        logger.error(output);
+      }
+
+      if (diagnostic.category === ts.DiagnosticCategory.Warning) {
+        logger.warn(output);
+      }
+
+      if (diagnostic.category === ts.DiagnosticCategory.Message) {
+        logger.info(output);
+      }
+
+      if (diagnostic.category === ts.DiagnosticCategory.Suggestion) {
+        logger.info(output);
+      }
+    } else {
+      logger.info(ts.flattenDiagnosticMessageText(diagnostic.messageText, '\n'));
+    }
+  }
+
+  if (emitResult.emitSkipped) {
+    const errors = allDiagnostics.filter((diag) => diag.category === ts.DiagnosticCategory.Error);
+
+    if (errors.length) {
+      throw new Error('Failed to compile TypeScript definitions');
+    }
+  }
+};
+
+/**
+ * @typedef {Object} DtsTaskEntry
+ * @property {string} exportPath
+ * @property {string} sourcePath
+ * @property {string} targetPath
+ */
+
+/**
+ * @typedef {Object} DtsTask
+ * @property {"build:dts"} type
+ * @property {DtsTaskEntry[]} entries
+ */
+
+/**
+ * @type {import('./index').TaskHandler<DtsTask>}
+ */
+const dtsTask = {
+  _spinner: null,
+  print(ctx, task) {
+    const entries = [
+      '   entries:',
+      ...task.entries.map((entry) =>
+        [
+          `    – `,
+          chalk.green(`${entry.importId} `),
+          `${chalk.cyan(entry.sourcePath)} ${chalk.gray('→')} ${chalk.cyan(entry.targetPath)}`,
+        ].join('')
+      ),
+      '',
+    ];
+
+    this._spinner = ora(`Building type files:\n`).start();
+
+    ctx.logger.log([...entries].join('\n'));
+  },
+  async run(ctx, task) {
+    await Promise.all(
+      task.entries.map(async (entry) => {
+        const config = await loadTsConfig({
+          /**
+           * TODO: this will not scale and assumes all project sourcePaths are `src/index.ts`
+           * so we can go back to the "root" of the project...
+           */
+          cwd: path.join(ctx.cwd, entry.sourcePath, '..', '..'),
+          path: 'tsconfig.dist.json',
```

Comment:
I'm wondering if having custom tsconfig names won't break the IDE parsers :thinking: 
I think I remember us having this issue when we first release the TS support with tsconfig.server.json & tsconfig.admin.json

Reply:
But maybe IDEs improved in the same time :man_shrugging:

Reply:
Firstly, we don't need IDEs to work with `tsconfig.dist.json` just like we dont need it to work with `tsconfig.eslint.json`. Secondly, talking from an FE perspective although realistically this will extend to BE – the base tsconfig _should_ include all the src files & your test files to ensure they're typed. But when your transpiling your sourcecode you don't need the test files, so we have a different `exclude` array. The other issue while you _could_ mitigate it with CLI arguments, is if you test your TS e.g `tsc` you typically want to use `noEmit = true` but we want declarations to be made by tsc in the build process, we also want to define the rootDir & outDir there as opposed to obfuscate it in the cli command.

This is the reasoning behind the dist file 😄 also we might want to start looking at API-Extractor for better documentation on our libraries and this incurs a few more options you don't really want for standard development imo.

Reply:
Hum ok the reasoning make sense. I would suggest renaming it to tsconfig.build.json to be consistant with the command name in that case